### PR TITLE
kubectl: fix deprecation warning bug

### PR DESCRIPTION
**Pull request by alexandercampbell**
**What this PR does / why we need it**:

Some kubectl commands were deprecated but would fail to print the
correct warning message when a flag was given before the command name.

	# Correctly prints the warning that "resize" is deprecated and
	# "scale" is now preferred.
	kubectl resize [...]

	# Should print the same warning but no warning is printed.
	kubectl --v=1 resize [...]

This was due to a fragile check on os.Args[1].

This commit implements a new function deprecatedCmd() that is used to
construct new "passthrough" commands which are marked as deprecated and
hidden.

Note that there is an existing "filters" system that may be preferable
to the system created in this commit. I'm not sure why the "filters"
array was not used for all deprecated commands in the first place.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by alexandercampbell** [comment]
@k8s-ci-robot signed CLA

**Comment by pwittrock** [comment]
clever

**Comment by pwittrock** [comment]
@k8s-bot ok to test

**Comment by pwittrock** [comment]
Looks good.  Are there existing tests for this?  If not, add some to https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test-cmd-util.sh

**Comment by alexandercampbell** [comment]
@pwittrock awesome, thanks. I was going to ask where tests for this should be written.

File: pkg/kubectl/cmd/cmd.go [hunk 1]
```diff
@@ -284,6 +284,7 @@ func NewKubectlCommand(f cmdutil.Factory, in io.Reader, out, err io.Writer) *cob
 				NewCmdCreate(f, out, err),
 				NewCmdExposeService(f, out),
 				NewCmdRun(f, in, out, err),
+				deprecatedCmd("run-container", NewCmdRun(f, in, out, err)),
```
Comment (droot) [inline pkg/kubectl/cmd/cmd.go]: sweet :)

**Comment by droot** [comment]
LGTM

**Review by droot — APPROVED** [review]

File: pkg/kubectl/cmd/cmd.go [hunk 2]
```diff
@@ -385,6 +391,20 @@ func printDeprecationWarning(command, alias string) {
...
+	// Have to be careful here because Cobra automatically extracts the name
+	// of the command from the .Use field.
+	originalName := cmd.Name()
+
+	cmd.Use = deprecatedVersion
+	cmd.Deprecated = fmt.Sprintf("use %q instead", originalName)
+	cmd.Hidden = true
+	return cmd
+}
+
```
Comment (droot) [inline pkg/kubectl/cmd/cmd.go]: was thinking more about the unit test for this function. We can write a unit test which wraps a NoOp cobra.Command and captures the "deprecated" string in the output. Thoughts ? (happy to discuss in person)

Comment (alexandercampbell) [inline pkg/kubectl/cmd/cmd.go]: Sounds good to me. Will look at that.

Comment (alexandercampbell) [inline pkg/kubectl/cmd/cmd.go]: Unit test written. Thanks for the suggestion!

**Comment by pwittrock** [comment]
@alexandercampbell Write a PTAL comment when you are ready to have us take another look

**Comment by alexandercampbell** [comment]
Thanks! I will do that as soon as I figure out what's up with the CI.

File: pkg/kubectl/cmd/cmd_test.go [hunk 3]
```diff
@@ -659,3 +662,38 @@ func genResponseWithJsonEncodedBody(bodyStruct interface{}) (*http.Response, err
...
+
+	if alias.Name() != "echo" {
+		t.Errorf("deprecatedAlias has name %q, expected %q",
+			alias.Name(), "echo")
+	}
+	if original.Name() != "print" {
+		t.Errorf("original command has name %q, expected %q",
+			original.Name(), "print")
+	}
+
```
Comment (droot) [inline pkg/kubectl/cmd/cmd_test.go]: hmm... I was thinking more along the lines of....invoking the deprecated command and then grabbing the output and check if deprecated warning is being emitted ?

**Review by droot — COMMENTED** [review]
Took a quick look at the unit test and have a few comments.

File: pkg/kubectl/cmd/cmd_test.go [hunk 4]
```diff
@@ -25,9 +25,12 @@ import (
 	"net/http"
 	"os"
 	"reflect"
+	stdstrings "strings"
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/cmd_test.go]: The name `strings` is taken by `k8s.io/kubernetes/pkg/util/strings` which is also imported in this package.

File: pkg/kubectl/cmd/cmd_test.go (hunk 3 above)
Comment (pwittrock) [inline pkg/kubectl/cmd/cmd_test.go]: We could also include the output writer as a dependency defaulted to stdout, but overridden the test, and see what bytes that it writes.

Comment (alexandercampbell) [inline pkg/kubectl/cmd/cmd_test.go]: I've updated the test to include more coverage-- see 5e1a3fd.

File: pkg/kubectl/cmd/cmd_test.go (hunk 4 above)
Comment (droot) [inline pkg/kubectl/cmd/cmd_test.go]: Got it. makes sense.

File: pkg/kubectl/cmd/cmd_test.go (hunk 3 above)
Comment (droot) [inline pkg/kubectl/cmd/cmd_test.go]: Just looked at it that commit. Looks good to me!

**Review by droot — COMMENTED** [review]
LGTM.

**Comment by alexandercampbell** [comment]
I tentatively believe that the tests are green.
PTAL, and let me know if you guys want me to rebase out the extraneous commits.

**Pull request merged**